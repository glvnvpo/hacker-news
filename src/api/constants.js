// @flow

type ID = number | string;

const BASE = 'https://hacker-news.firebaseio.com/v0';

export const NEW_STORIES = () => `${BASE}/newstories.json`;
export const ITEM = (id: ID)=> `${BASE}/item/${id}.json`;