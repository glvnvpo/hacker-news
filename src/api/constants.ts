// @flow

import {ID, URL} from '../types';

const BASE = 'https://hacker-news.firebaseio.com/v0';

export const NEW_STORIES = (): URL => `${BASE}/newstories.json`;
export const ITEM = (id: ID): URL => `${BASE}/item/${id}.json`;