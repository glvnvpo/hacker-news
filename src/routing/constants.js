export const ROOT_PATH                 = '/';
export const MAIN_PAGE_PATH            = '/hacker-news';
export const SINGLE_STORY_PAGE_PATH    = `${MAIN_PAGE_PATH}/:id`;