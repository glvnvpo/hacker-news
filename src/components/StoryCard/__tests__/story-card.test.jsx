import React from 'react';
import {mount} from 'enzyme';
import {BrowserRouter} from "react-router-dom";
import {StoryCard} from "../index";
import {MAIN_PAGE_PATH} from "../../../routing/constants";

describe('StoryCard', () => {
	let wrapper;

	const story = {
		id: 123,
		title: 'someTitle',
		by: 'someUser',
		time: '1662661417',
		score: 2
	};
    
	it('should render correct StoryCard', () => {
		const props = {
			story,
			fieldsToShow: ['id', 'title', 'by', 'time', 'score'],
			asLink: true,
			to: `${MAIN_PAGE_PATH}/${story.id}`
		};
		wrapper = mount(
			<BrowserRouter>
				<StoryCard {...props} />
			</BrowserRouter>
		);
		expect(wrapper).toMatchSnapshot();
	});

	it('should render correct StoryCard with all props', () => {
		const props = {
			story: {
				...story,
				kids: [],
				text: "someText",
				url: "https://reactjs.org/",
				descendants: 0
			}
		};
		wrapper = mount(<StoryCard {...props} />);
		expect(wrapper).toMatchSnapshot();
	});

	it('should render without url', () => {
		const props = {
			story: {
				...story,
				kids: [],
				text: "someText",
				descendants: 0
			}
		};
		wrapper = mount(<StoryCard {...props} />);
		expect(wrapper.find('a')).toHaveLength(0);
	});

	it('should render without text', () => {
		const props = {
			story: {
				...story,
				kids: [],
				descendants: 0
			}
		};
		const expected = "To read the text, visit the source";
		wrapper = mount(<StoryCard {...props} />);
		expect(wrapper.text()).toMatch(expected);
	});
	
	it('should render story with error', () => {
		wrapper = mount(<StoryCard story={{}} />);
		const expected = "Some troubles in loading story";
		expect(wrapper.text()).toBe(expected);
	});

	it('should render Spinner', () => {
		wrapper = mount(<StoryCard isLoading={true} story={{}} />);
		expect(wrapper.find('Spinner')).toHaveLength(1);
	});
});