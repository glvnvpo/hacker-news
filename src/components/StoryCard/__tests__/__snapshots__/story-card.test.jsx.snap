// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryCard should render correct StoryCard 1`] = `
<BrowserRouter>
  <Router
    location={
      Object {
        "hash": "",
        "key": "default",
        "pathname": "/",
        "search": "",
        "state": null,
      }
    }
    navigationType="POP"
    navigator={
      Object {
        "action": "POP",
        "createHref": [Function],
        "go": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "default",
          "pathname": "/",
          "search": "",
          "state": null,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <StoryCard
      asLink={true}
      fieldsToShow={
        Array [
          "id",
          "title",
          "by",
          "time",
          "score",
        ]
      }
      story={
        Object {
          "by": "someUser",
          "id": 123,
          "score": 2,
          "time": "1662661417",
          "title": "someTitle",
        }
      }
      to="/hacker-news/123"
    >
      <Card
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          }
        }
        body={false}
        className="story mt-10 mb-20"
        to="/hacker-news/123"
      >
        <Link
          className="story mt-10 mb-20 card"
          to="/hacker-news/123"
        >
          <a
            className="story mt-10 mb-20 card"
            href="/hacker-news/123"
            onClick={[Function]}
          >
            <CardBody>
              <div
                className="card-body"
              >
                <CardTitle
                  className="color-orange"
                >
                  <ForwardRef
                    className="color-orange card-title"
                  >
                    <div
                      className="color-orange card-title h5"
                    >
                      someTitle
                    </div>
                  </ForwardRef>
                </CardTitle>
                <CardSubtitle
                  className="mb-2 color-grey"
                >
                  <ForwardRef
                    className="mb-2 color-grey card-subtitle"
                  >
                    <div
                      className="mb-2 color-grey card-subtitle h6"
                    >
                      2
                       points | 
                      someUser
                       | 
                      08.09.2022
                    </div>
                  </ForwardRef>
                </CardSubtitle>
              </div>
            </CardBody>
          </a>
        </Link>
      </Card>
    </StoryCard>
  </Router>
</BrowserRouter>
`;

exports[`StoryCard should render correct StoryCard with all props 1`] = `
<StoryCard
  story={
    Object {
      "by": "someUser",
      "descendants": 0,
      "id": 123,
      "kids": Array [],
      "score": 2,
      "text": "someText",
      "time": "1662661417",
      "title": "someTitle",
      "url": "https://reactjs.org/",
    }
  }
>
  <Card
    body={false}
    className="story mt-10 mb-20"
  >
    <div
      className="story mt-10 mb-20 card"
    >
      <CardBody>
        <div
          className="card-body"
        >
          <CardTitle
            className="color-orange"
          >
            <ForwardRef
              className="color-orange card-title"
            >
              <div
                className="color-orange card-title h5"
              >
                someTitle
              </div>
            </ForwardRef>
          </CardTitle>
          <CardSubtitle
            className="mb-2 color-grey"
          >
            <ForwardRef
              className="mb-2 color-grey card-subtitle"
            >
              <div
                className="mb-2 color-grey card-subtitle h6"
              >
                2
                 points | 
                someUser
                 | 
                08.09.2022
              </div>
            </ForwardRef>
          </CardSubtitle>
          <CardSubtitle
            className="mb-2 color-grey"
          >
            <ForwardRef
              className="mb-2 color-grey card-subtitle"
            >
              <div
                className="mb-2 color-grey card-subtitle h6"
              >
                <a
                  href="https://reactjs.org/"
                  rel="noreferrer"
                  target="_blank"
                >
                  Source
                </a>
              </div>
            </ForwardRef>
          </CardSubtitle>
          <CardSubtitle
            className="mb-2 color-grey"
          >
            <ForwardRef
              className="mb-2 color-grey card-subtitle"
            >
              <div
                className="mb-2 color-grey card-subtitle h6"
              >
                Comment count: 
                0
              </div>
            </ForwardRef>
          </CardSubtitle>
          <CardText
            as="span"
            className="bold color-dark-grey"
          >
            <span
              className="bold color-dark-grey card-text"
            >
              someText
            </span>
          </CardText>
        </div>
      </CardBody>
    </div>
  </Card>
</StoryCard>
`;
